MINI-PROJECTS.md

```diff
@@ -104,6 +104,314 @@ share a bound.
+multiple SAT solvers but no command line options to pick a SAT solver. It would be
+nice if there were.
+
+César is working on this one. (?)
+
+### Combine Uninitialize Variables Analysis with Alias Analysis
+
+`goto-instrument` features an instrumentation to track cases of use-before-define, i.e.
+reading of uninitialised variables. At present, the analysis does not take into account
+aliasing information. As such, the following code will yield a spurious counterexample:
+
+``` C
+int main()
+{
+  int x;
+  int *ptr;
+  ptr=&x;
+  *ptr=42;
+  if(x < 42)
+    return 1;
+  return 0;
+}
+```
+
+### Constant Folding
+
+With the recently added reaching definitions analysis, it is straightforward to implement
+a program transformation that performs constant folding. ~To ensure soundness in concurrent
+settings, it should be combined with the existing is_threaded analysis.~ (A concurrency-aware
+version of reaching definitions now exists and is awaiting a merge.) Along the way, assertions
+that become trivially true should be eliminated.
+
+### Make SAMATE Tests Work
+
+Read [this paper](https://www.sciencedirect.com/science/article/abs/pii/S0950584913000384) and
+make the tests mentioned in there work.
+
+### Error Explanation
+
+Re-implement the algorithm described in [this paper](http://www.kroening.com/papers/STTT-error-explanation-2005.pdf)
+
+### LTL
+
+Build an instrumenter in `goto-instrument` for checking LTL (or even CTL) properties on C code.
+E.g look at [this paper](https://www.umsec.umn.edu/publications/Partial-Translation-Verification-Untrusted-Code)
+(And please get in touch with Michael Tautschnig, who has done very similar work in FShell).
+
+Rajdeep is now looking into this one. (?)
+
+### Test SMT Back Ends
+
+CBMC allows to choose among a selection of SMT solvers to use as solving back end. As these
+tools evolve, CBMC's code for running them may need a fresh look from time to time. After
+a first round of manual experiments, these tests should become part of the routine build
+process on `dkr-build`. (Michael Tautschnig knows how to do this.)
+
+### Simple Test Harness Generation
+
+CBMC currently outputs a trace of execution when it find a counter-example. Instead it
+should be possible to produce a C file that when compiled and run will cause the program
+to execute this trace. This will allow a faster verify/debug cycle. (To a certain
+extent this has been accomplished in FShell. Speak to Michael.)
+
+Pascal is now looking at this one. (?)
+
+### Improved Interpreter
+
+CProver has an interpreter for goto programs but it could be improved, for example adding
+support for dynamic analysis, calls to external functions and possibly even mixed concrete
+and symbolic execution. It would also be good to be able to interpret traces found by BMC
+to catch mismatches / bugs between the logical model and the bit blasting models. This
+would also be useful for more advanced CEGAR like techniques (also see below).
```

Comment:
I have no idea what the actual state of the interpreter is today.

Reply:
Using the abstract interpreter would also be possible.